temporal-sdk/src/main/java/io/temporal/internal/worker/AsyncActivityPollTask.java

```diff
@@ -0,0 +1,141 @@
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AsyncActivityPollTask implements AsyncPoller.PollTaskAsync<ActivityTask> {
+  private static final Logger log = LoggerFactory.getLogger(AsyncActivityPollTask.class);
+
+  private final TrackingSlotSupplier<?> slotSupplier;
+  private final WorkflowServiceStubs service;
+  private final Scope metricsScope;
+  private final PollActivityTaskQueueRequest pollRequest;
+  private final AtomicInteger pollGauge = new AtomicInteger();
+  private final Context.CancellableContext grpcContext = Context.ROOT.withCancellation();
+
+  @SuppressWarnings("deprecation")
+  public AsyncActivityPollTask(
+      @Nonnull WorkflowServiceStubs service,
+      @Nonnull String namespace,
+      @Nonnull String taskQueue,
+      @Nonnull String identity,
+      @Nullable String buildId,
+      boolean useBuildIdForVersioning,
+      double activitiesPerSecond,
+      @Nonnull TrackingSlotSupplier<ActivitySlotInfo> slotSupplier,
+      @Nonnull Scope metricsScope,
+      @Nonnull Supplier<GetSystemInfoResponse.Capabilities> serverCapabilities) {
+    this.service = service;
+    this.slotSupplier = slotSupplier;
+    this.metricsScope = metricsScope;
+
+    PollActivityTaskQueueRequest.Builder pollRequest =
+        PollActivityTaskQueueRequest.newBuilder()
+            .setNamespace(namespace)
+            .setIdentity(identity)
+            .setTaskQueue(TaskQueue.newBuilder().setName(taskQueue));
+    if (activitiesPerSecond > 0) {
+      pollRequest.setTaskQueueMetadata(
+          TaskQueueMetadata.newBuilder()
+              .setMaxTasksPerSecond(DoubleValue.newBuilder().setValue(activitiesPerSecond).build())
+              .build());
+    }
+
+    if (serverCapabilities.get().getBuildIdBasedVersioning()) {
+      pollRequest.setWorkerVersionCapabilities(
+          WorkerVersionCapabilities.newBuilder()
+              .setBuildId(buildId)
+              .setUseVersioning(useBuildIdForVersioning)
+              .build());
+    }
+    this.pollRequest = pollRequest.build();
+  }
+
+  @Override
+  public CompletableFuture<ActivityTask> poll(SlotPermit permit) {
+    if (log.isTraceEnabled()) {
+      log.trace("poll request begin: " + pollRequest);
+    }
+
+    MetricsTag.tagged(metricsScope, PollerTypeMetricsTag.PollerType.ACTIVITY_TASK)
+        .gauge(MetricsType.NUM_POLLERS)
+        .update(pollGauge.incrementAndGet());
+
+    CompletableFuture<PollActivityTaskQueueResponse> response = null;
+    try {
+      response =
+          grpcContext.call(
+              () ->
+                  GrpcUtils.toCompletableFuture(
+                      service
+                          .futureStub()
+                          .withOption(METRICS_TAGS_CALL_OPTIONS_KEY, metricsScope)
+                          .pollActivityTaskQueue(pollRequest)));
+    } catch (Exception e) {
+      throw new RuntimeException(e);
```

Comment:
This might also need to decrement poll gauge? (applies to all poll types, as well as the deployment options comment)

Reply:
I am not sure `call` will ever actually throw in my use case, but you are right we should decrement here regardless